### ISSUE TITLE
Make SRCCACHE usage more consistent

### ISSUE THREAD

**Issue by staticfloat** [JuliaLang/julia#31530]
The purpose behind `$(SRCCACHE)` is that it should be an immutable cache of pristine source tarballs.  We really shouldn't be patching these things, because otherwise when we change our patches, we can run into weird situations that require us to clean out the source cache and re-download everything.  Instead, we should use `$(BUILDDIR)` as the place where things get patched, and `$(SRCCACHE)` should only ever be pristine things that do not change from julia build to julia build.  Yes, for things that build out of tree this will necessitate us dealing with three locations: one for source download, one for source unpacking/patching, one for build outputs.  I suggest using the convention `$(SRCCACHE)` for downloading, `$(BUILDDIR)/<name>` for unpacking/patching, `$(BUILDDIR)/<name>/build` for out-of-tree builds.

Incomplete list of things known to disobey this rule (note that I am asserting that we should not even extract source into `$(SRCCACHE)` but should instead extract into `$(BUILDDIR)`:

* [LLVM](https://github.com/JuliaLang/julia/blob/d60503a00edfa19d6b13f8aba8ff3e0e4cdf9ae0/deps/llvm-options.mk#L23)
* [CURL](https://github.com/JuliaLang/julia/blob/d60503a00edfa19d6b13f8aba8ff3e0e4cdf9ae0/deps/curl.mk#L24)
* [GMP](https://github.com/JuliaLang/julia/blob/d60503a00edfa19d6b13f8aba8ff3e0e4cdf9ae0/deps/gmp.mk#L16)

**Comment by vtjnash** [comment]
How about we use `$(SRCDIR)/tarballs`? Or maybe we should add secondary cache logic to the jldownloads script? On posix systems, we might be able to make hardlinks for each tarballs from $new_download_location to $SRCDIR to ease the transition.

**Comment by staticfloat** [comment]
> How about we use $(SRCDIR)/tarballs?

Use that for what?  A new default location for `$(SRCCACHE)`?

> Or maybe we should add secondary cache logic to the jldownloads script?

What would this logic do?

> On posix systems, we might be able to make hardlinks for each tarballs from $new_download_location to $SRCDIR to ease the transition.

I don't really want to change where the tarballs get stored; I want to change where they get unpacked to.

**Comment by vtjnash** [comment]
> What would this logic do?

Bypass the call to network IO in favor of making a hardlink from a cache location.

**Comment by vtjnash** [comment]
> I don't really want to change where the tarballs get stored; I want to change where they get unpacked to.

But other than who gets to stay in the same place, what difference does it make? It's much easier to move tarballs than to move source directories.

**Comment by staticfloat** [comment]
> But other than who gets to stay in the same place, what difference does it make?

Ah; it's because `$(SRCCACHE)` persists between builds (because it gets overridden and placed within a persistent location) whereas everything within the julia build tree gets wiped.  I don't really care where things are, just that the tarballs and the unpacked directories are in disjoint locations so that I can store one within the julia git repository folder (such that `git clean -fdx` will clear it) and the tarballs can be placed outside.

> Bypass the call to network IO in favor of making a hardlink from a cache location.

Don't we already have that?  We don't do much network I/O on the buildbots right now because `$(SRCCACHE)` is persistent.  What would the "secondary cache" do?

**Comment by vtjnash** [comment]
A secondary cache would be even more persistent probably.

I guess what you want, iiuc, is a separate variable for the source (extract?) from the tarballs (get?), but that both default to `deps/srccache`? That seems workable.

**Comment by staticfloat** [comment]
> I guess what you want, iiuc, is a separate variable for the source (extract?) from the tarballs (get?), but that both default to `deps/srccache`? That seems workable.

Yes, that is pretty much what I want.  I don't actually care where the extraction target defaults to (I was thinking `deps/scratch`, not `deps/srccache`) but as long as its configurable it makes me happy.

**Comment by vtjnash** [comment]
Btw, `rm -rf srccache/*/` along with `git clean -fdx` would kill all folders in the cache too, although I foresee that could still limit the ability to do parallelism, if that is also a concern

**Comment by staticfloat** [comment]
> Btw, rm -rf srccache/*/ along with git clean -fdx would kill all folders in the cache too, although I foresee that could still limit the ability to do parallelism, if that is also a concern

Yes, but I'm trying to avoid hitting the network, so I want the tarballs to persist but not the patched, extracted, source trees.

**Comment by ViralBShah** [comment]
Close?

**Comment by inkydragon** [comment]
Do we still want to extract and patch files in `$(BUILDDIR)`?


Now there are 14 dependencies that will be extracted into the `$(SRCCACHE)` folder.

<details>
<summary>full list</summary>

- curl
- gmp
- libgit2
- libssh2
- libunwind
- libuv
- llvm
- llvmunwind
- mbedtls
- mpfr
- nghttp2
- patchelf
- pcre2
- zlib

</details>
